analytics/pubstack/eventchannel/channel.go

```diff
@@ -0,0 +1,127 @@
+package eventchannel
+
+import (
+	"bytes"
+	"compress/gzip"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"path"
+	"syscall"
+	"time"
+
+	"github.com/golang/glog"
+)
+
+func (c *Channel) resetMetrics() {
+	c.metrics.eventCount = 0
+	c.metrics.bufferSize = 0
+	c.metrics.eventError = 0
+}
+
+type ChannelMetrics struct {
+	bufferSize int64
+	eventCount int64
+	eventError int64
+}
+
+type Channel struct {
+	intake  *url.URL
+	gz      *gzip.Writer
+	buff    *bytes.Buffer
+	ch      chan []byte
+	metrics ChannelMetrics
+}
+
+// Add : add a new event to be processed
+func (c *Channel) Add(event []byte) {
+	c.ch <- event
+}
+
+func (c *Channel) forward(maxSize, maxCount int64, maxTime time.Duration, termCh chan os.Signal) {
+	ticker := time.NewTicker(maxTime)
+
+	for {
+		select {
+		// termination received
+		case <-termCh:
+			glog.Info("termination signal received")
+			c.flush()
+			return
+		// event is received
+		case event := <-c.ch:
+			_, err := c.gz.Write(event)
+			if err != nil {
+				c.metrics.eventError++
+				glog.Warning("fail to compress event")
+				continue
+			}
+			c.metrics.eventCount++
+			c.metrics.bufferSize = int64(c.buff.Len())
+			if c.metrics.eventCount >= maxCount || c.metrics.bufferSize >= maxSize {
+				c.flush()
+			}
+		// time between flushes has passed
+		case <-ticker.C:
+			c.flush()
+		}
+	}
+}
+
+func (c *Channel) flush() {
+	c.resetMetrics()
```

Comment:
I'm concerned of a race condition between flushing and writing new data. Did you run tests with race detection turned on?

Reply:
I have reworked that part. I'll appreciate if you can look again here.
The race detection only shows only an issue with the thread initializer, like all adapters, I think we are good there.